### ISSUE TITLE
Add basic Proxy support for non-supported platforms

### ISSUE THREAD

**Issue by RReverser** [tvcutsem/harmony-reflect#50]
Would it be possible to add at least traps for `Object.*` static methods on platforms that do not support real proxies?

I'm currently playing with some code where at least such basic support would be useful, and from my understanding it shouldn't be hard to implement it since overrides for those methods are already implemented and just need some basic `Proxy` implementation for storing and accessing handlers.

**Comment by tvcutsem** [comment]
At one point I had exactly this idea, but I never pursued it because such
emulated Proxies would provide only a half-baked API, whose limitations
would need to be properly documented. Also, I don't see a real use case for
an API that only manipulates objects via the Object.* methods and never via
property access (dot), for-in loops, the `delete` operator, etc.

If your use case for manipulating an object only with Object.* methods is
sufficiently appealing, then maybe.

Bottom line: I don't like the cost/benefit tradeoff of this feature request
because the use case is narrow and the patch is unlikely to be small (given
it needs proper documentation + tests).

2015-05-21 11:58 GMT+02:00 Ingvar Stepanyan <notifications@github.com>:

**Comment by RReverser** [comment]
> that only manipulates objects via the Object.* methods

`apply` and `construct` can be emulated, too (just return new function which calls target one on certain conditions)

This way only `deleteProperty`, `enumerate`, `get`, `has` and `set` will be left as non-polyfillable.

Moreover - if we accept deviations from spec (which would be unfortunate, but probably better than not polyfilling at all) we could use `enumerate` on Proxy creating (instead of `for-in` as in real Proxy), get the list of keys and return objects with such keys and ES5 `get`/`set` handlers. This way `enumerate`, `get`, `set` and `has` traps will be available at least partially and only `deleteProperty` will be left.

**Comment by tvcutsem** [comment]
`construct` can be emulated, but only imperfectly (can't distinguish `new F()` from `F.call(Object.create(F.prototype))`) I just rediscovered issue #8 from years ago where the same ideas were brought up.

The problem with using the properties of the `target` object at creation time is that we don't have a way of updating the properties on the proxy if new props are added or existing props are deleted from `target`.

The above 2 examples are partly what I meant by "half-baked API" and "requires extensive documentation to describe the limitations".

Overall I think a polyfill for Proxy on non-Proxy supporting browsers may be better implemented as a separate project. With this shim I want to stay as close to the spec as possible, and I don't want to frustrate users by having them think they can use the full Proxy API reliably in non-Proxy supporting browsers.

I don't expect to be able to start a separate Proxy polyfill project myself soon, but always happy to help out with testing & reviewing code if anyone else is up for it.

**Comment by RReverser** [comment]
> I don't expect to be able to start a separate Proxy polyfill project myself soon
Ok, no problem, I can work on it on my own.

> but always happy to help out with testing & reviewing code if anyone else is up for it
That will be highly appreciable, I'll let you know when it's ready.

Just one question here - would you mind if I would "steal" some pieces like static `Object.*` methods and probably some tests from your implementation?

**Comment by tvcutsem** [comment]
Not at all. The code of this project is dual-licensed under Mozilla MPL and Apache License. You're free to reuse whatever is needed to get the polyfill working.

When the project's ready, I would be happy to include a link to it in the README.

**Comment by RReverser** [comment]
Looking at https://github.com/tvcutsem/harmony-reflect/blob/master/reflect.js#L1432 - shouldn't be `Object.defineProperties` patched as well?

**Comment by tvcutsem** [comment]
You're right. The built-in version of `defineProperties` will not call the patched `Object.defineProperty` for each individual property. I'll open a separate issue for this. Thanks!

**Comment by RReverser** [comment]
Btw, ` // Patches:` section in comments seems to be out-of-date (does not include `defineProperty` / `defineProperties` and maybe others).
